synchronizing with code (#1273)
elixir-lang/elixir-lang.github.com: getting-started/mix-otp/supervisor-and-application.markdown

{% include mix-otp-preface.html %}

> **TODO**: Please state what is the use of reading this page, what we're going to learn, how we're going to use it, what problem it solves.

Reviewer: What do you want to say here that we do not talk about in the next 3 paragraphs?

Author: as I said: what's the goal of the page. the first paragraph sounds like a very generic introduction to the need for recovery from error. second paragraph introduces the term 'supervisor' (lower case), that "will start a fresh copy of our registry if a bug leads it to crash" (the two parts of the sentence are miles apart). third paragraph introduces the term 'application' (again lower case), next to supervisor (still lower case) and it states we are going to enhance our program with **two** supervisors, without explaining: (1) what is a supervisor, (2) why do we need two? and what about this 'application' thing?  and why aren't we talking of the `Task`, I was expecting here? or is this `Supervisor` some sort of smarter or specialised `GenServer`? all questions roaming in my head while trying to make sense of the text.

it's only after this (for me confusing) introduction that we finally see the `Supervisor` base module we are going to `use` in our `KV.Supervisor`, and before I know what will be the role of this new thing, I see code and I'm being explained of what the coding details mean.

Author: ah, ok, I see the text you added, in the next section, it makes sense, I would probably want it in the introduction.

can you help me understand the relation `GenServer` / `Supervisor`?  possibly using some parallels with OO terms?

Reviewer: I am not sure we can answer all of your concerns. Part of an introduction is to provide an overall description of what we are going to build. For you to learn exactly what a supervisor is, you will continue reading. :)

## Our first supervisor

> **TODO**: state the role of this first supervisor, possibly relating it to the role of the second supervisor, which was also not mentioned.

Reviewer: ```suggestion
A supervisor is a process which supervises other processes, which we refer to as child processes. The act of supervising a process includes three distinct responsibilities. The first one is to start child processes. Once a child process is running, the supervisor may restart a child process, either because it terminated abnormally or because a certain condition has reached. For example, a supervisor may restart all children if any child dies. Finally, a supervisor is also responsible for shutting down the child processes on the system is shutting down. Please see the [Supervisor](https://hexdocs.pm/elixir/Supervisor.html) module for a more in-depth discussion.
```

> Creating a supervisor is not much different from creating a GenServer. We are going to define a module named `KV.Supervisor`, which will use the [Supervisor](https://hexdocs.pm/elixir/Supervisor.html) behaviour, inside the `lib/kv/supervisor.ex` file:

Reviewer: ```suggestion
Creating a supervisor is not much different from creating a GenServer. We are going to define a module named `KV.Supervisor`, which will use the Supervisor behaviour, inside the `lib/kv/supervisor.ex` file:
```

end
```

> > Please note that by doing this, we are breaking the boilerplate test case which defined our Application as a hello world greeter. We could simply remove that test case.

Reviewer: ```suggestion
> Please note that by doing this, we are breaking the boilerplate test case which defined our Application as a hello world greeter. You can simply remove that test case.
```
